Add blog post for 1.29 (#373)
kubewarden/kubewarden.io: content/blog/2025/09/kubewarden-1.29-release.md

---
title: "Kubewarden 1.29 Release"
authors:
  - Víctor Cuadrado Juan
date: 2025-09-30
---

Straight from the kitchen, Kubewarden 1.29 is served! This release is a medley
bowl of healthy stack features, crisp policy improvements, and some fresh

> fixes, all seasoned with the wholesome flavour of paid-off tech debt.

Reviewer: Well humored first paragraphs can be a new standard to our release blog post! :) 

I'm just afraid that I'll not be able to write them well as you do!

## Removal of Picky dependency and stringent behavior change

Since long we have been depending on the Rust crate `picky` as the
implementation for X.509 and PKI certificates that we use in [our cryptographic host
capabilities](https://docs.kubewarden.io/reference/spec/host-capabilities/crypto).
It allowed us to overcome some limitations in the `webpki` crate.

These limitations have been resolved recently in the `rustls-wepbki` crate, a
fork of the original `webpki` project. For us, this meant that we could perform
the library migration as part of our recurring tech-debt efforts.

This library change from `picky` to `rustls-wepbki` for the cryptographic
host capability brings stricter behaviour when validating certificates:

- Key usage: Previously, certificates were checked only against signing usage.
  Now, we accept any key usage, which opens more use cases.
- Change with certificate chains: Now, a certificate will be considered

>   not trusted if its intermediate CA is expired. Wepki does not check the

Reviewer 2: ```suggestion
  untrusted if its intermediate CA is expired. Wepki does not check the
```

  if no chain is provided, the Mozilla's CA is used.

This host capability is exposed in our [cel
policy](https://artifacthub.io/packages/kubewarden/cel-policy/cel-policy), and
also in our [verify-image-signatures](https://artifacthub.io/packages/kubewarden/verify-image-signatures/verify-image-signatures)
policy (point 5 of the readme).

## Granular timeout protection for policies

Policies usually run and finish fast enough. Yet there can be some that perform
more complex validations: for example the
[`env-variable-secrets-scanner`](https://artifacthub.io/packages/kubewarden/env-variable-secrets-scanner/env-variable-secrets-scanner)
policy, which uses a Rust library to scan for secrets in environment variables.
For these policies, it is beneficial to extend the evaluation time alloted to
them.

Up until now, this was possible globally per PolicyServer by setting their env
var `KUBEWARDEN_POLICY_TIMEOUT` via its `spec.env`.

With 1.29.0, we have added a new field to all policy resources:
`spec.timeoutEvalSeconds`.

This field complements the already existing `spec.timeoutSeconds`, which sets
the timeout for the associated Kubernetes Webhhok for the policy.

> The new `spec.timeoutEvalSeconds` field is optional and allows to specify the

Reviewer 2: ```suggestion
The new `spec.timeoutEvalSeconds` field is optional and allows specifying the
```

  - slow
```

This new `allowedStorageClasses` setting cannot be used in conjunction with
`deniedStorageClasses` (you either have an allow-list, or a deny-list). One
can always deploy several policies, though.

## ContainerDays Hamburg 2025

One of our maintainers (Víctor, your narrator today!) had a speaker slot
on [ContainerDays 2025
Hamburg](https://www.containerdays.io/containerdays-conference-2025), with the
talk _"Simplifying cluster security with CEL"_. In it we gave an overview on
the CEL language, its uses in both Kubernetes Admission Controllers and Dynamic
Admission Controllers such as Kubewarden, and explained how we expanded on it
for Kubewarden.

Stay tuned to their [YouTube channel](https://www.youtube.com/@ContainerDays)
to watch the recording!

## Google Summer of Code wrapped up!

We are happy to congratulate Esosa Ohangbon, our Google Summer Of Code
mentee, on successfully completing the mentorship! 🎉🥳🎊

Esosa worked on fleshing out a new [Policy SDK for
Typescript/Javascript](https://www.npmjs.com/package/kubewarden-policy-sdk).
The end result is a tech-preview of a useful SDK. We look forward to
collaborations such as this one!

## Signing Hauler manifest

> Since Kubewarden 1.28, we also ship a [Hauler
> manifest](../08/kubewarden-1.28-release/) together with our Helm charts to
> simplify air-gap installs.

Reviewer 3: ```suggestion
Since [Kubewarden 1.28](../08/kubewarden-1.28-release/), we also ship a [Hauler
manifest](https://docs.hauler.dev/docs/intro) together with our Helm charts to
simplify air-gap installs.
```

With 1.29, we now cryptographically sign the Hauler manifest, the image and the
policies list.

For example, to verify with Cosign the `hauler_manifest.yaml`:

```console
$ cosign verify-blob --bundle hauler_manifest_cosign.bundle \
  --certificate-identity-regexp 'https://github.com/kubewarden/*' \
  --certificate-oidc-issuer https://token.actions.githubusercontent.com \
  hauler_manifest.yaml
Verified OK
```

## Fixed Crypto host capability regression

On Kubewarden 1.28, we shipped a minor feature to allow performing session
replays of the [Crypto host capability](https://docs.kubewarden.io/reference/spec/host-capabilities/crypto).
This host capability is the one used in only one of the possible settings for
the verify-image-signatures policy: verifying via certificates ([point 5 in its
readme](https://github.com/kubewarden/verify-image-signatures)).

> Doing so, we broke this specific host capability API, therefore, this policy
> setting too.

Reviewer 2: ```suggestion
In doing so, we broke this specific host capability API, and therefore, this policy
setting too.
```
